### PR TITLE
Add BaseProgram example. Support 'approve' and 'revoke' instructions …

### PR DESCRIPTION
**Pull request by czl1378**
…for SPLToken. #1

### CODE REVIEW

File: src/SPLToken.ts [hunk 1]
```diff
@@ -353,6 +382,68 @@ export class SPLToken extends BaseProgram {
...
+
+    const layout = BufferLayout.struct([BufferLayout.u8('instruction')]);
+
+    let keys: Array<any> = [{ write: account }];
+
+    if (multiSigners.length === 0) {
+      keys.push(revokeAuthority);
+    } else {
+      multiSigners.forEach(signer => keys.push(signer));
+    }
```
Comment (defactojob) [inline src/SPLToken.ts:442]: maybe we could extend `InstructionAuthority` to consider `Account[]` or `PublicKey[]` as a multisigners.

we see the multisigners pattern in many token instructions.

then this method boils down to:

```ts
this.instructionEncode(layout, { instruction: 5 }, [
   revokeAuthority || multiSigners
]);
```

File: src/SPLToken.ts [hunk 2]
```diff
@@ -206,21 +226,30 @@ export class SPLToken extends BaseProgram {
...
       BufferLayout.u8('instruction'),
       uint64('amount'),
     ])
 
+    let keys: Array<any> = [
+      { write: token },
+      { write: to }
+    ]
+
+    if (multiSigners.length === 0) {
```
Comment (defactojob) [inline src/SPLToken.ts:242]: see comment below, for an idea how to abstract multiSigners better.

File: src/SPLToken.ts [hunk 3]
```diff
@@ -54,13 +54,33 @@ interface MintToParams {
...
   mintAuthority: Account
+  multiSigners: Array<Account>
+}
+
+interface ApproveParams {
+  account: PublicKey
+  delegate: PublicKey
+  amount: bigint
+  approveAuthority: Account,
+  multiSigners: Array<Account>,
```
Comment (defactojob) [inline src/SPLToken.ts:66]: let's use `T[]` to declare array types

File: src/BaseProgram.ts [hunk 4]
```diff
@@ -38,8 +38,11 @@ export abstract class BaseProgram {
     for (let inst of insts) {
       tx.add(inst)
     }
-
-    return await sendAndConfirmTransaction(this.conn, tx, signers)
+   
+    return await sendAndConfirmTransaction(this.conn, tx, signers, {
+      commitment: 'singleGossip',
+      preflightCommitment: 'singleGossip',
```
Comment (defactojob) [inline src/BaseProgram.ts:44]: instead of hardwiring `singleGossip`, we could read these from this.conn (I believe)

File: src/BPFLoader.ts [hunk 5]
```diff
@@ -4,21 +4,21 @@ import {
...
   public async load(programBinary: Buffer, programAccount = new Account()): Promise<Account> {
     await BpfLoader.load(
       this.wallet.conn,
       this.wallet.account,
       programAccount,
       programBinary,
-      this.programID,
-    )
+      BPF_LOADER_PROGRAM_ID,
+    );
```
Comment (defactojob) [inline src/BPFLoader.ts:20]: maybe we add a linter to standardize ; or not

**Review by defactojob — COMMENTED** [review]
looks good

i think we refine the multisigner API

File: src/index.ts [hunk 6]
```diff
@@ -8,13 +8,15 @@ export { BaseProgram } from "./BaseProgram"
...
-export type NetworkName = "dev" | "main"
+export type NetworkName = "local" | "dev" | "main"
 
 export namespace solana {
   export function connect(networkName: NetworkName, opts: { commitment?: string } = {}): Connection {
     const commitment = opts.commitment || "singleGossip"
 
     switch (networkName) {
+      case "local":
+        return new Connection("http://localhost:8899", commitment as any)
```
Comment (defactojob) [inline src/index.ts:19]: how about let's change the API to:

```
connect("dev")
connect("http://localhost:8899")
```

and "local" aliases to "http://localhost:8899"

**Review by defactojob — CHANGES_REQUESTED** [review]

File: src/BaseProgram.ts (hunk 4 above)
Comment (czl1378) [reply src/BaseProgram.ts:44]: After some testing, I found that if we do not specify  `preflightCommitment` or `skipPreflight` here,  will cause the error: `Transaction simulation failed: Attempt to load a program that does not exist `.

Comment (defactojob) [reply src/BaseProgram.ts:44]: i mean you can change it to:

```
commitment: this.conn.commitment,
preflightCommitment: this.conn.commitment,
```

File: src/SPLToken.ts (hunk 1 above)
Comment (czl1378) [reply src/SPLToken.ts:442]: good idea, i will consider to implement this

File: src/BaseProgram.ts (hunk 4 above)
Comment (czl1378) [reply src/BaseProgram.ts:44]: Okay.

File: src/index.ts (hunk 6 above)
Comment (czl1378) [reply src/index.ts:19]: maybe next update.

File: src/index.ts [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
-export { BaseProgram } from "./BaseProgram"
-export { BPFLoader } from "./BPFLoader"
-export { System } from "./System"
+export { Account, PublicKey } from "@solana/web3.js"
+export { default as Wallet } from "./Wallet"
+export { default as SPLToken } from "./SPLToken"
+export { default as ProgramAccount } from "./ProgramAccount"
+export { default as BaseProgram } from "./BaseProgram"
+export { default as BPFLoader } from "./BPFLoader"
+export { default as System } from "./System"
```
Comment (defactojob) [inline src/index.ts:9]: let's ban default exports. sorry.

File: src/SPLToken/index.ts [hunk 8]
```diff
@@ -0,0 +1,463 @@
+import { TransactionInstruction, SYSVAR_RENT_PUBKEY } from '@solana/web3.js';
```
Comment (defactojob) [inline src/SPLToken/index.ts:1]: please move `SPLToken/index.ts -> SPLToken.ts`.

impossible to code review now.

File: src/SPLToken/types.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  owner: PublicKey
+  account?: Account
+}
+
+export interface MintToParams {
+  token: PublicKey
+  to: PublicKey
+  amount: bigint
+  authority: Account | PublicKey
+  multiSigners: Account[]
```
Comment (defactojob) [inline src/SPLToken/types.ts:38]: I think it'd be more concise to just combine these together.

```
authority: Account | Account[]
```

The authority cannot be PublicKey, but it needs to be a signer.

**Review by defactojob — CHANGES_REQUESTED** [review]

**Review by defactojob — APPROVED** [review]

**Pull request merged**